README.md
# 网站快速开发模板
申明：本开发模板适用于小型快速简单网站开发，非用于大型或者模块化开发之使用。

## 开发环境
采用 Nodejs 为基础的 gulp 搭建前端自动化模块，并利用 npm 工具对前端库进行管理。通过以下命令全局安装 gulp：
```
npm install -g gulp-cli
```

## 安装依赖包
将代码 clone 下来后，首次运行之前，需要需安装项目依赖包。通过以下命令，我们将开发环境的自动化工具以及网站所需要的一些第三方框架都初始化好。这些初始的包和框架列表写在 `package.json` 中。
```
npm install
```

在开发过程中，如果需要其他的第三方前端框架，也通过以下命令来安装并引用：
```
npm install -D package_name
npm install -S package_name
```

## 项目启动
在 gulpfile.js 中已经配置好项目启动后的一些初始化工作，所以只要简单的执行以下命令即可：
```
gulp
```

如果是手机网站，在浏览器模拟器中，可以从地址 `http://localhost:9000/`进入，项目跟目录下的页面模拟了iOS safari浏览器的状态栏、地址栏和工具栏，是你能够更加清楚的知道网页在浏览器中的页面尺寸。

## HTML 文件生成
本项目采用 gulp-file-include 插件，静态 HTML 代码将通过 include 标记插入到指定的 html 文件中，然后生成完整 html 文件转移到 `app` 目录中。

## SCSS 文件说明
通过 gulp 任务实时编译 css 文件到 `app/css/` 目录下。

## JS 文件说明
已全面支持 ES6 语法，通过 gulp 任务实时编译成 ES5 代码，然后存放到 `app/js/` 下。

### 自建 minxins

| Mixins | 说明 |
| --- | --- |
| `size(width, height)` | 方便的宽高设定，如只传一个参数，则长宽一样 |
| `clearfix` | 清除浮动|
| `absolute(argu)` | 绝对定位，按需传入定位参数即可，eg: `@include absolute(top 0 right 0)`|
| `fixed(argu)` | 固定定位，同上|
| `relative(argu)` | 相对定位，同上|
| `respond-to(argu)` | 响应式断点设置，断点设置见 `_variables.scss`|

### 自建 functions
以下是一些 SCSS 目前还不支持的数学运算：

| 函数 | 说明 |
| --- | --- |
| `pow(number, exp)`| 乘方 |
| `fact(number)`    | 阶乘 |
| `rad(angle)`      | 角度转弧度 |
| `sin(angle)`      | 正弦 |
| `cos(angle)`      | 余弦 |
| `tan(angle)`      | 正切 |

以上 function 依项目需求而定，可不引入。在 docs.scss 中取消 import 即可。

## 测试说明

### JS 测试
~~已在 gulpfile.js 中 配置了 task。在命令行中执行 `gulp jshint` 即可进行 JS 语法测试。~~

现已通过 eslint 来对 es6 代码进行同步测试，不用 gulp task 了。

### SCSS 测试
已在 gulpfile.js 中 配置了 task。在命令行中执行 `gulp scsslint` 即可进行 SCSS 语法测试。
```
gulp scsslint
```

说明：因 gulp-scss-lint 语法检查模块依赖 Ruby 环境下的 scss-lint 包，所以在 Windows 环境下不是太方便，因此，该测试亦可量力而行，有环境可以执行以下命令后再运行该 task 进行测试。
```
gem install scss-lint
```

## 项目打包
`gulp build` 任务会将所有 css 和 js 都合并、压缩以及混淆后，输出到 dist 目录。 可以在 `http://localhost:9000/dist/` 中访问发布出来的文件用以检测。
```
gulp build
```

检验完毕后，这里也提供了一个压缩打包的 task 供使用。执行 `gulp zip` 命令，会在项目根目录下生成一个 `achive_xxxx.zip` 文件，可以将该文件提交给相关发布人员发布到正式环境中。
```
gulp zip
```


CHANGELOG.md
# CHANGE LOG

## 2020-04-11
- 依赖包升级

## 2019-07-10
- 升级到 gulp 4，重写了所有 gulp 任务。
- 同时升级了其他的一些依赖包。

## 2017-02-13
- 进一步优化 gulp 文件，精简 task 数量。
- 升级了一系列 npm 包的版本，特别是 gulp-sass 的版本，避免了 node-sass 不兼容当前 node 版本的问题(v7.4.0)
- 排查了之前项目打包会产生 MaxListenersExceededWarning 的情况，屏蔽了 gulp-imagemin 产生该 warning 的报错信息（[详情](https://github.com/sindresorhus/gulp-imagemin/issues/237)）
- 修改了 gulp-sass 任务中默认 sass 路径的配置错误
- 引入 gulp-sequence，定制 gulp task 的执行顺序
- 提升版本号为 _2.3.5_

## 2017-01-06
- 重新优化打包流程，分离资源提取与压缩步骤，大大节约了 build 时间。

## 2016-10-17
- 改用 gulp-file-include 来生成静态页面。
- 将 gulp-minify-css 替换为 gulp-clean-css。
- 修改 .eslint 文件，使其适应于 jQuery 的语法检测。

## 2016-09-27
众多更新，版本号升级为 _2.0.0_

- scss 采用 BS4 alpha 4 的 reboot
- 使用 es2015 来编写 js，添加新 task 实时编译成 ES5
- gulpfile.js 也采用 es2015 编写，保留 ES5 版本以防万一（gulpfile_es5.js）
- 调整目录结构，src 目录用于存放css，js 和 html 的预编译源码

## 2016-09-22
升级项目内的 npm 包，升级后，某些依赖包的 API 发生变化，重写 gulpfile.js。

## 2015-09-28
scss 框架更改，reset 样式的代码从 Bootstrap 4 中挖过来占为己有，同时提升版本号：_0.9.0_

## 2015-08-06
多项重大更新：

* 调整了项目的目录结构，规定 app 目录下为无需编译或者已编译文件
* 采用 gulp-file-wrapper 插件编译静态文件
* 采用 browser sync 替换 connect 组件
* 删除 sandal 样式框架，建立新的 scss 架构
* 更新 README 文档
* 更新版本号为： _0.8.6_
* 添加 CHANGELOG 文档

## 2015-06-04
* 添加 scsslint，为规范 scss 文件做 test 校验

## 2015-05-29
* 更新 npm 包，升级为更新版本，提高稳定性

## 2015-05-28
* 为 template 文件添加 watch
* 更新 README, 增加 sandal 框架及自身的 scss 使用说明
* 更新项目版本号

## 2015-05-27
* image task
* 抛弃 ruby-sass，采用 node-sass 引擎，提升 scss 文件编译效率
* 采用 gulp-file-include 渲染静态文件

## 2015-05-04
* 创建项目


gulpfile.js
/**
 * gulp command
 */
const { src, dest, task, watch, parallel, series } = require('gulp')
/**
 * gulp plugin
 */
const [ scssLint, scssLintStylish, fileInclude, sourcemaps, sass, autoprefixer, babel, useref, htmlmin, uglify, cleanCss, zip, del, browserSync ] = [
  require('gulp-scss-lint'),
  require('gulp-scss-lint-stylish'),
  require('gulp-file-include'),
  require('gulp-sourcemaps'),
  require('gulp-sass'),
  require('gulp-autoprefixer'),
  require('gulp-babel'),
  require('gulp-useref'),
  require('gulp-htmlmin'),
  require('gulp-uglify'),
  require('gulp-clean-css'),
  require('gulp-zip'),
  require('del'),
  require('browser-sync').create()
]
/**
 * 文件路径设置
 */
const _ = {
  app: 'app', dist: 'dist',
  js: 'app/js', css: 'app/css', img: 'app/img',
  scss: 'src/scss', tmpl: 'src/pages', es6: 'src/es6',
  assets: 'src/assets'
}
/**
 * BrowserSync notify style
 */
const styles = [
  'display: block',
  'padding: 15px',
  'font-family: sans-serif',
  'position: fixed',
  'font-size: 0.9em',
  'z-index: 9999',
  'bottom: 0px',
  'right: 0px',
  'border-top-left-radius: 5px',
  'background-color: rgba(0,0,0,0.4)',
  'margin: 0',
  'color: white',
  'text-align: center'
]
/**
 * Scss lint
 */
task('scsslint', () => src(`${_.scss}/**/*.scss`)
  .pipe(scssLint({
    'config': '.scsslintrc',
    'customReport': scssLintStylish
  }))
)
/**
 * Render html file to static
 */
task('html', () => src(`${_.tmpl}/*.html`)
  .pipe(fileInclude({
    prefix: '@@',
    basepath: '@file'
  }))
  .pipe(dest(`${_.app}/`))
)
/**
 * SCSS files compiler, autoprefixer as post-css
 */
task('scss', () => src(`${_.scss}/**/*.scss`)
  .pipe(sourcemaps.init())
  .pipe(sass({
    outputStyle: 'expanded',
    sourceComments: false
  }))
  .pipe(autoprefixer({
    Browserslist: ['> 1%']
  }))
  .pipe(sourcemaps.write('./'))
  .pipe(dest(_.css))
  .pipe(browserSync.stream())
)
/**
 * ES6 files compiler
 */
task('es6', () => src(`${_.es6}/**/*.es6`)
  .pipe(sourcemaps.init())
  .pipe(babel({
    presets: ['@babel/env']
  }))
  .pipe(sourcemaps.write('./'))
  .pipe(dest(_.js))
)
/**
 *  Pickup images and move to app/img folder
 */
task('opt-img', () => src(`${_.assets}/**/*.{png,jpg,jpeg,gif,ico}`)
  .pipe(dest(_.img))
)

// Build Task
/**
 * Move images to dist folder
 */
task('image', () => src(`${_.img}/**/*`)
  .pipe(dest(`${_.dist}/img/`))
)
/**
 * Comb CSS files and JS files to dist folder
 */
task('assets', () => src(`${_.app}/*.html`)
  .pipe(useref())
  .pipe(dest(_.dist))
)
/**
 * HTML files minify
 */
task('html-minify', () => src(`${_.dist}/*.html`)
  .pipe(htmlmin({
    removeComments: true,
    collapseWhitespace: true
  }))
  .pipe(dest(_.dist))
)
/**
 * JS files minify
 */
task('js-minify', () => src(`${_.dist}/js/*.js`)
  .pipe(uglify())
  .pipe(dest(`${_.dist}/js`))
)
/**
 * CSS files minify
 */
task('css-minify', () => src(`${_.dist}/css/*.css`)
  .pipe(cleanCss())
  .pipe(dest(`${_.dist}/css`))
)
/**
 * Zip dist files
 */
task('zip', () => src(`${_.dist}/**/*`)
  .pipe(zip(`archive_${Date.now()}.zip`))
  .pipe(dest('./'))
)
/**
 * Watch files
 */
task('watch', () => {
  // Watch scss files
  watch(`${_.scss}/**/*.scss`, parallel('scss'))
  // Watch es6 files
  watch(`${_.es6}/**/*.es6`, parallel('es6'))
  // Watch template files
  watch(`${_.tmpl}/**/*.html`, parallel('html'))
  // Watch image files
  watch(`${_.assets}/**/*.{png,jpg,jpeg,gif,ico}`, parallel('opt-img'))

  // Watch for reload
  watch([
    `${_.js}/**/*.js`,
    `${_.app}/*.html`,
    `${_.img}/**/*.{png,jpg,jpeg,gif,ico}`
  ]).on('change', browserSync.reload)
})

// Utility Task
/**
 * Clean folder
 */
task('clean-app', () => del([_.app]))
task('clean-dist', () => del([_.dist]))
/**
 * Run a server and open project
 */
task('runner', () => {
  browserSync.init({
    ui: false,
    server: { baseDir: './' },
    startPath: './app',
    port: 9000,
    notify: { styles }
  })
})
/**
 * Start the proje (main task)
 */
task('start', series('clean-app', 'scss', 'es6', 'html', 'opt-img', 'runner'))

exports.default = parallel('start', 'watch')
/**
* Build files to dist
*/
exports.build = series('clean-app', 'scss', 'es6', 'html', 'opt-img', 'clean-dist', 'assets', 'image', 'html-minify', 'js-minify', 'css-minify')


gulpfile3.js
(function(gulp, gulpLoadPlugins) {
  let $, _, styles, taskInit

  /* gulp 插件自动载入 */
  $ = gulpLoadPlugins({
    pattern: ['*'],
    rename: { 'jshint': 'Jshint'}
  })

  /* 文件路径设置 */
  _ = {
    app: 'app', dist: 'dist',
    js: 'app/js', css: 'app/css', img: 'app/img',
    scss: 'src/scss', tmpl: 'src/pages',  es6: 'src/es6'
  }

  /* BrowserSync 通知样式 */
  styles = [
    'display: none',
    'padding: 15px',
    'font-family: sans-serif',
    'position: fixed',
    'font-size: 0.9em',
    'z-index: 9999',
    'bottom: 0px',
    'right: 0px',
    'border-top-left-radius: 5px',
    'background-color: rgba(0,0,0,0.4)',
    'margin: 0',
    'color: white',
    'text-align: center'
  ]

  /* 项目启动初始任务 */
  taskInit = ['scss', 'es6', 'html']

  /* 启动多浏览器同步 webserver */
  gulp.task('bs', taskInit, () => {
    $.browserSync.init({
      ui: false,
      server: { baseDir: './' },
      startPath: './app',
      port: 9000,
      notify: { styles }
    })
  })

  /* JS 文件格式校验（非 es6） */
  gulp.task('jshint', () => gulp
    .src(['gulpfile.js', `${_.js}/**/*.js`])
    .pipe($.jshint())
    .pipe($.jshint.reporter($.jshintStylish))
  )

  /* scss 文件格式校验 */
  gulp.task('scsslint', () => gulp
    .src([`${_.scss}/**/*.scss`, `!${_.scss}/utils/*.scss`])
    .pipe($.scssLint({
      'config': '.scsslintrc',
      'customReport': $.scssLintStylish
    }))
  )

  /**
   编译预处理代码，供预览用
  */
  /* html 代码片段渲染成静态 HTML 文件 */
  gulp.task('html', () => gulp
    .src(`${_.tmpl}/*.html`)
    .pipe($.plumber())
    .pipe($.fileInclude({
      prefix: '@@',
      basepath: '@file'
    }))
    .pipe(gulp.dest(`${_.app}/`))
  )

  /* scss 文件编译成普通 CSS，添加 autoprefixer 后处理 */
  gulp.task('scss', () => gulp
    .src(`${_.scss}/**/*.scss`)
    .pipe($.plumber())
    .pipe($.sourcemaps.init())
    .pipe($.sass({
      outputStyle: 'expanded',
      sourceComments: false,
      includePaths: ['./bower_components/bootstrap/scss/']
    }).on('error', $.sass.logError))
    .pipe($.autoprefixer({
      browsers: ['last 15 versions', '> 1%']
    }))
    .pipe($.sourcemaps.write('./'))
    .pipe(gulp.dest(_.css))
    .pipe($.browserSync.stream())
  )

  /* es6 文件编译成普通 es5文件 */
  gulp.task('es6', () => gulp
    .src(`${_.es6}/**/*.es6`)
    .pipe($.sourcemaps.init())
    .pipe($.babel({
      presets: ['es2015', 'stage-2']
    }))
    .pipe($.sourcemaps.write('./'))
    .pipe(gulp.dest(_.js))
  )

  /**
   打包任务
  */
  /* images 优化 */
  gulp.task('image', () => gulp
    .src(`${_.img}/**/*`)
    .pipe($.imagemin([
      $.imagemin.gifsicle(),
      $.imagemin.jpegtran(),
      // $.imagemin.optipng(),
      $.imagemin.svgo()
    ]))
    .pipe(gulp.dest(`${_.dist}/img/`))
  )

  /* 分离 HTML 中需要要再处理的 JS 和 CSS 文件，并按设置合并 */
  gulp.task('assets', () => gulp
    .src(`${_.app}/*.html`)
    .pipe($.plumber())
    .pipe($.useref())
    .pipe(gulp.dest(_.dist))
  )

  /* HTML 文件压缩 */
  gulp.task('html-minify', () => gulp
    .src(`${_.dist}/*.html`)
    .pipe($.plumber())
    .pipe($.htmlmin({
      removeComments: true,
      collapseWhitespace: true
    }))
    .pipe(gulp.dest(_.dist))
  )

  /*  js 文件混淆及压缩 */
  gulp.task('js-minify', () => gulp
    .src(`${_.dist}/js/*.js`)
    .pipe($.plumber())
    .pipe($.uglify())
    .pipe(gulp.dest(`${_.dist}/js`))
  )

  /*  CSS 文件压缩 */
  gulp.task('css-minify', () => gulp
    .src(`${_.dist}/css/*.css`)
    .pipe($.plumber())
    .pipe($.cleanCss())
    .pipe(gulp.dest(`${_.dist}/css`))
  )

  /* 将 build 好的文件压缩成 zip 文件，并打上时间戳 */
  gulp.task('zip', () => gulp
    .src(`${_.dist}/**/*`)
    .pipe($.plumber())
    .pipe($.zip(`archive_${Date.now()}.zip`))
    .pipe(gulp.dest('./'))
  )

  /* watch task */
  gulp.task('watch', () => {
    // Watch scss files
    $.watch(`${_.scss}/**/*.scss`, { ignoreInitial: false }, () => gulp.start('scss'))
    // Watch es6 files
    $.watch(`${_.es6}/**/*.es6`,   { ignoreInitial: false }, () => gulp.start('es6'))
    // Watch template files
    $.watch(`${_.tmpl}/**/*.html`, { ignoreInitial: false }, () => gulp.start('html'))

    // Watch for reload
    $.watch([
      `${_.js}/**/*.js`,
      `${_.app}/*.html`,
      `${_.img}/**/*.{png,jpg,jpeg,gif,ico}`
    ], $.browserSync.reload)
  })

  /*  dist 目录清除 */
  gulp.task('clean', () => $.del([_.dist]))

  /* 发布任务定制 */
  gulp.task('build', $.sequence(
    taskInit, 'clean', 'assets',
    ['image', 'html-minify', 'js-minify', 'css-minify'])
  )

  /* 默认任务，运行项目 */
  gulp.task('default', ['bs', 'watch'])
}(require('gulp'), require('gulp-load-plugins')))


gulpfile_es5.js
(function(gulp, gulpLoadPlugins) {
  'use strict';
  var $, _, notifyStyle;

  // gulp autoload plugins
  $ = gulpLoadPlugins({
        pattern: ['*'],
         rename: { 'jshint': 'Jshint'}
      });

  // path setting
  _ = {
    app:  'app',      dist: 'dist',
    js:   'app/js',   css:  'app/css',    img:  'app/img',
    scss: 'src/scss', tmpl: 'src/pages',  es6:  'src/es6'
  };

  // BrowserSync notify style
  notifyStyle = [
    'display: none',
    'padding: 15px',
    'font-family: sans-serif',
    'position: fixed',
    'font-size: 0.9em',
    'z-index: 9999',
    'bottom: 0px',
    'right: 0px',
    'border-top-left-radius: 5px',
    'background-color: rgba(0,0,0,0.4)',
    'margin: 0',
    'color: white',
    'text-align: center'
  ];

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ Wait for scss build & html templates render, then launch the Server
  //|~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('bs', ['scss', 'es6', 'static'], function() {
    $.browserSync.init({
      ui: false,
      server: { baseDir: './' },
      startPath: './app',
      port: 9000,
      notify: { styles: notifyStyle}
    });
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ jshint - js files test
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('jshint', function() {
    return gulp.src([ 'gulpfile.js',  _.js + '/**/*.js'])
      .pipe($.jshint())
      .pipe($.jshint.reporter($.jshintStylish));
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ scsslint - scss files test
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('scsslint', function() {
    return gulp.src([
      _.scss + '/**/*.scss', '!' + _.scss + '/utils/*.scss'
      ])
      .pipe($.scssLint({
        'config': '.scsslintrc',
        'customReport': $.scssLintStylish
      }));
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ render template from html file
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('static', function() {
    return gulp.src([_.tmpl + '/partial/**/*.html'])
      .pipe($.plumber())
      .pipe($.fileWrapper(_.tmpl + '/layout.html'))
      .pipe(gulp.dest(_.app + '/'))
      .pipe($.size({
        title: 'HTML files:'
      }));
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ scss2css (use node-sass)
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('scss', function() {
    return gulp.src(_.scss + '/**/*.scss')
      .pipe($.plumber())
      .pipe($.sourcemaps.init())
      .pipe($.sass({
        outputStyle: 'expanded'
      }).on('error', $.sass.logError))
      .pipe($.autoprefixer({
         browsers: ['last 15 versions', '> 1%']
       }))
      .pipe($.sourcemaps.write('./'))
      .pipe(gulp.dest(_.css))
      .pipe($.browserSync.stream())
      .pipe($.size({
        title: 'CSS files:'
      }));
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ es6 => es5 (use present-es2015)
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('es6', function() {
    return gulp.src(_.es6 + '/**/*.js')
      .pipe($.sourcemaps.init())
      .pipe($.babel({
          presets: ['es2015']
      }))
      .pipe($.sourcemaps.write('./'))
      .pipe(gulp.dest(_.js))
      .pipe($.size({
        title: 'JS files:'
      }));
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ optimize images
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('image', ['clean'], function () {
    return gulp.src(_.img + '/**/*')
      .pipe($.imagemin({
        progressive: true
      }))
      .pipe(gulp.dest(_.dist + '/img/'))
      .pipe($.size({
        title: 'IMAGE files:'
      }));
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ join & minify css & js
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('assets', ['clean'], function() {
    return gulp.src(_.app + '/*.html')
      .pipe($.plumber())
      .pipe($.useref())
      .pipe($.if('*.js', $.uglify()))
      .pipe($.if('*.css', $.minifyCss({
        keepSpecialComments: 0
      })))
      .pipe(gulp.dest(_.dist));
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ minify html files
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('minify', ['assets'], function() {
    return gulp.src(_.dist + '/*.html')
      .pipe($.plumber())
      .pipe($.htmlmin({
        removeComments: true,
        collapseWhitespace: true
      }))
      .pipe(gulp.dest(_.dist));
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ zip dist files with time stamp
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('zip', function() {
    return gulp.src(_.dist + '/**/*')
      .pipe($.plumber())
      .pipe($.zip('archive_' + (+new Date()) + '.zip'))
      .pipe(gulp.dest('./'))
      .pipe($.size({
        title: 'ZIP file:'
      }));
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ watch
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('watch', function () {
      // Watch scss files
      gulp.watch(_.scss + '/**/*.scss', ['scss']);
      // Watch es6 files
      gulp.watch(_.es6 + '/**/*.js', ['es6']);
      // Watch template files
      gulp.watch([_.tmpl + '/**/*.html'], ['static']);
      gulp.watch([
        _.js  + '/**/*.js',
        _.app + '/*.html',
        _.img + '/**/*.{png,jpg,jpeg,gif,ico}'
      ]).on('change', $.browserSync.reload);
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ~ clean dist folder
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('clean', function() {
    return $.del([_.dist]);
  });

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ✓ alias
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  // gulp.task('test',  ['jshint', 'scsslint']);
  gulp.task('build', ['image', 'minify']);

  //|**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  //| ✓ default
  //'~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
  gulp.task('default', ['bs', 'watch']);

}(require('gulp'), require('gulp-load-plugins')));
